services/supervisor/start_agent.go

```diff
@@ -80,39 +89,64 @@ func (sup *SupervisorService) getContainerUnsafe(name string) (*Container, bool)
 
 func (sup *SupervisorService) addContainerUnsafe(container *clients.DockerContainer, agentConfig ...*config.AgentConfig) {
 	if agentConfig != nil {
-		sup.containers = append(sup.containers, &Container{
-			DockerContainer: *container,
-			IsAgent:         true,
-			AgentConfig:     agentConfig[0],
-		})
+		sup.containers = append(
+			sup.containers, &Container{
+				DockerContainer: *container,
+				IsAgent:         true,
+				AgentConfig:     agentConfig[0],
+			},
+		)
 		return
 	}
 	sup.containers = append(sup.containers, &Container{DockerContainer: *container})
 }
 
 func (sup *SupervisorService) handleAgentRun(payload messaging.AgentPayload) error {
+	startCtx, cancel := context.WithTimeout(sup.ctx, agentStartTimeout)
+	defer cancel()
+
+	return sup.handleAgentRunWithContext(startCtx, payload)
+}
+
+func (sup *SupervisorService) handleAgentRunWithContext(ctx context.Context, payload messaging.AgentPayload) error {
 	sup.lastRun.Set()
 
-	log.WithFields(log.Fields{
-		"payload": len(payload),
-	}).Infof("handle agent run")
+	log.WithFields(
+		log.Fields{
+			"payload": len(payload),
+		},
+	).Infof("handle agent run")
+
+	var wg sync.WaitGroup
+
+	wg.Add(len(payload))
 
 	for _, agent := range payload {
-		err := sup.startAgent(agent)
-		if err == errAgentAlreadyRunning {
-			log.Infof("agent container '%s' is already running - skipped", agent.ContainerName())
-			sup.msgClient.Publish(messaging.SubjectAgentsStatusRunning, messaging.AgentPayload{agent})
-			continue
-		}
-		if err != nil {
-			log.Errorf("failed to start agent: %v", err)
-			continue
-		}
+		go sup.doStartAgent(ctx, agent, &wg)
+	}
+
+	wg.Wait()
+
+	return nil
+}
+
+// doStartAgent intended to use during multiple agent starts
+func (sup *SupervisorService) doStartAgent(ctx context.Context, agent config.AgentConfig, wg *sync.WaitGroup) {
+	defer wg.Done()
 
-		// Broadcast the agent status.
+	err := sup.startAgent(ctx, agent)
+	if err == errAgentAlreadyRunning {
+		log.Infof("agent container '%s' is already running - skipped", agent.ContainerName())
 		sup.msgClient.Publish(messaging.SubjectAgentsStatusRunning, messaging.AgentPayload{agent})
+		return
 	}
-	return nil
+	if err != nil {
+		log.Errorf("failed to start agent: %v", err)
```

Comment:
one thing we started doing in the rest of the system is this kind of thing:
```
logger := log.WithFields(...agentID info...etc)
...
logger.WithError(err)
```
so that the log can be parsed/structured and include relevant data (like which agent is failing)

Reply:
makes sense, added. (eyeballed the fields)